split pending variants into mempool and transmitted

AloeareV opened the pull request.

fluidvanadium commented on zingolib/src/wallet/transaction_context.rs:
```diff
@@ -458,6 +458,7 @@ mod decrypt_transaction {
                             note.clone(),
                             to,
                             output_index,
+                            status.is_mempool(),
```
why not just change add_pending_note to use status as an argument? replaci

AloeareV replied:
Well...this would allow passing a `Confirmed` status to add_pending_note, which we would then need to explicitly error-handle.

fluidvanadium replied:
pub(crate) add_pending_note is called only inside account_for_shielded_receipts, which takes status as a parameter, immediately after a status check

if there is shared behavior between add_new_note and add pending note, they must be merged. the merged function would receive a confirmation_status and have a conditional if status.is_pending() blockclause. 

breaking and returning Err if the status is not pending is functionally identical to the above

zancas replied:
This is marked outdated by github.

fluidvanadium commented on zingolib/src/wallet/data.rs:
```diff
@@ -1738,8 +1738,11 @@ pub mod summaries {
+        TransmittedSpent(TxId),
+        /// Output is pending spent.
+        /// The transaction consuming this output has been detected in the mempool.
+        MempoolSpent(TxId),
```
I believe this is @zancas 's wet code. If variants of SpendSummary correspond to to variants of ConfirmationStatus, use an enum wrapper!

AloeareV replied:
We have a method that creates one from an  `Option<(Txid, ConfirmationStatus)>`. We could have it be defined as a newtype wrapper (or enum wrapper, like `enum SpendSummary { Unspent, SpendStatus(ConfirmationStatus)}`, I do think it's a tradeoff. There's a readability benefit to having the definition show all the possible variants at the top level of the enum, especially where they're all have their own variant documentation. 

Edit: The more I think about it the less I'm convinced that it's much of a tradeoff...I very well might make that change while I'm looking at tests

zancas replied:
This is marked outdated by github.

fluidvanadium reviewed (dismissed):
I requested some changes

fluidvanadium commented:
This is great. I would merge it as is! I left comments in the way of nearby nodes in the infinite map of further upgrades

AloeareV committed:
fix polarity of assertions

AloeareV committed:
Merge branch 'dev' into split_pending_mempool_sent

AloeareV committed:
make mempool tests run, assert transactions are now mempool...attempt to fix to match expectation

AloeareV committed:
attempt to figure out test hang

fluidvanadium reviewed (changes requested):
after the most recent two commits, i am confused. why did this PR grow instead of merging?

fluidvanadium commented:
im gonna grab the best version and run with it

AloeareV commented:
> after the most recent two commits, i am confused. why did this PR grow instead of merging?

There were no running tests that tested mempool functionality, that I could tell. Turned that on, changed some tests, fixed a bug that the tests caught, but the tests are still failing.

AloeareV committed:
fixed hang by dropping lock when needed

AloeareV committed:
canary test now passes. todo: cleanup and rest of test suite

AloeareV committed:
fix test

AloeareV committed:
Merge remote-tracking branch 'origin/dev' into split_pending_mempool_sent

AloeareV committed:
fix FIXME test

AloeareV committed:
remove extra dbgs

AloeareV committed:
clippy fix

AloeareV committed:
fmt

AloeareV commented:
Not sure what's going on with CI. These tests all pass locally

AloeareV committed:
increase sleep to maybe help ci

AloeareV committed:
make ci feature that turns off mempool tests

fluidvanadium committed:
Merge branch 'dev' into split_pending_mempool_sent

zancas committed:
Merge branch 'dev' into split_pending_mempool_sent

zancas committed:
the feature is sync, the crate is zingo-sync

zancas commented on zingolib/src/lightclient/sync.rs:
```diff
@@ -191,34 +191,52 @@ impl LightClient {
+                                    if matches!(r.status, ConfirmationStatus::Transmitted(_)) {
+                                        // In this case, we need write access, to change the status
+                                        // from Transmitted to Mempool
+                                        drop(tms_readlock);
```
Nice.
